File: 1886-minimum-limit-of-balls-in-a-bag/1886-minimum-limit-of-balls-in-a-bag.c

int minimumSize(int* nums, int numsSize, int maxOperations) {
    int i, left, right = 0, mid;

    for (i = 0, left = 0; i < numsSize; i++)
        if (left < nums[i])
            left = nums[i];

    while (left > right) {
        mid = (left + right) / 2 + (left + right) % 2;

        long long count = 0;

        for (i = 0; i < numsSize; i++)
            count += nums[i] / mid - 1 + (nums[i] % mid > 0);

        if (count > maxOperations)
            right = mid;
        else
            left = mid - 1;
    }

    return left + 1;
}

File: 1886-minimum-limit-of-balls-in-a-bag/1886-minimum-limit-of-balls-in-a-bag.go
func minimumSize(nums []int, maxOperations int) int {
    l, r := 1, 0
    maximum_num := func (x, y int) int { 
        if x > y { 
            return x;
        }; 
        return y; 
    }

    for _, num := range nums {
        r = maximum_num(r, num)
    }

    solve := func(nums []int, cost int, maxOperations int) bool {
        cnt := 0
        for _, v := range nums {
            cnt += (v - 1) / cost
            if cnt > maxOperations { 
                return false
            }
        }
        return true
    }
    for l <= r {
        m := (l + r) >> 1
        if solve(nums, m, maxOperations) {
            r = m - 1
        } else {
            l = m + 1
        }
    }
    return l
}